### DISCUSSION TITLE
持续会话疑问

### DISCUSSION THREAD

**Discussion by YOLO-MIC** [simple-robot/Discussions#184] [Q&A] [answered]
### 涉及的编程语言

Java

### 涉及的组件库

mirai (QQ)

### 疑问描述

前因：我想搞明白持续会话如何使用，在看了例子(https://github.com/orgs/simple-robot/discussions/122)，
试下来发现，在我用A账号发送给bot“改名字”，用B账号发送给bot“”Mike”时，bot仍会给回复A账号改名成功。
之后我又找文档想知道如何针对同一账号进行持续会话时看到下面这段
[![image](https://github.com/simple-robot/Discussions/assets/93651899/065537e5-5a02-46dc-8a9b-ab2c9f45706d)
图中下面的那段话是说使用next这个API可以快速判断对话是否是同一个人发送的消息的意思吗？试下来似乎并不是这样，所以请问有方便的api可以快速对同一个好友展开持续会话吗？

**Comment by ForteScarlet** [comment]
我记得 next 需要一个 `Event` 类型的参数，它的作用就是根据这个 event 的类型来替你完成一些基础的判断，比如相同的id之类的

**Reply by ForteScarlet** [reply]
这段文档的下面一点的地方会有一个它的判断逻辑的示例
![image](https://github.com/simple-robot/Discussions/assets/40045247/3a437e9e-cdea-4284-9e3c-723de6a28b3c)

**Reply by YOLO-MIC** [reply]
嗯 这个我都看了  我想说如果要实现判断是否是同一个好友  还是要和我问题里那个图里一样做判断吗？

**Reply by ForteScarlet** [reply]
正常来讲，应该将好友消息那个 event 放进 next 就可以了，如果这样子不行的话，那可以暂时先用你的那个图里一样做逻辑判断，有可能是个小bug之类的

**Reply by YOLO-MIC** [reply]
我的版本是3.3.0  是直接拿[这里](https://github.com/orgs/simple-robot/discussions/122)的例子试的

**Reply by ForteScarlet** [reply, marked as answer]
如果也是使用的 `nextMessage(event, FriendMessageEvent.Key, ...);` ，但是没有生效的话，那么改成 `waitingForNextMessage` 然后自行处理判断逻辑试试看

**Reply by YOLO-MIC** [reply]
试过了，这样是可以的

**Reply by ForteScarlet** [reply]
那就先暂时这样吧，3.x的持续会话的反馈量很少，可能会存在一些bug或者缺陷😟
_但是按理说 #122 里能成功，那应该就可以才对.. 很奇怪_

**Reply by YOLO-MIC** [reply]
好的，有可能他没试过用不同号也能触发同一个持续会话？我自己封装先这样用着了，感谢

**Reply by YOLO-MIC** [reply]
你好，还有个问题，如果我想bot主动进行持续会话，我应该如何获取`ContinuousSessionContext`这个对象，我看持续会话文档里需要`EventProcessingContext `或者`EventListenerProcessingContext`,应该如何获取呢

**Reply by ForteScarlet** [reply]
持续会话对象存于此处：
https://github.com/simple-robot/simpler-robot/blob/bf4b1e14a2cb42af3a6b8acd9450a7a007b6c1d9/simbot-core/src/main/kotlin/love/forte/simbot/core/event/SimpleEventListenerManager.kt#L84-L100
可以先获取 `Application`，然后获取其中的 `eventListenerManager`，并判断类型是否为 `SimpleEventListenerManager`, 如果是，则转化类型并获取即可

**Reply by YOLO-MIC** [reply]
ok，感谢

**Reply by YOLO-MIC** [reply]
你好，又有一个疑问，如何实现同一个QQ只能开启一个持续会话，在尝试开启第二个持续会话时先关闭上一个会话？
场景描述：在群里进行一段持续会话，
A->Bot: 改名字
Bot->A: 改什么名字（名字只能是数字 使用 `waitingForNext`过滤数字）
A->Bot: 改名字 （这时候又尝试开启一个会话，我想要先将上一次的会话关闭该如何实现？）
在V2版本中可以使用 `ContinuousSessionScopeContext`里的 `get`和 `remove`方法实现，这个版本里有类似的操作吗？

**Reply by ForteScarlet** [reply]
可以参考一下 provider 和 receiver https://simple-robot-library.github.io/simbot3-website/docs/basic/continuous-session#%E6%8C%81%E7%BB%AD%E4%BC%9A%E8%AF%9D 且创建会话的时候应该也可以提供一个 id，可以用来事后获取，provider 我记得应该是可以被取消掉的

**Reply by YOLO-MIC** [reply]
好的我先尝试一下

**Reply by YOLO-MIC** [reply]
ok，尝试了一下，这样大致是可以的
```java
@Listener
    @Filter("test")
    public void demo2(GroupMessageEvent event, ContinuousSessionContext continuousSessionContext) {
        ContinuousSessionProvider<Object> provider = continuousSessionContext.getProvider("aaa");
        if (provider != null) {
            provider.cancel(new RobotException("cancel last session"));
        }
        event.replyBlocking("输入一个数字：");
        GroupMessageEvent event2 = continuousSessionContext.waitingForNext("aaa", GroupMessageEvent.Key, (context, event3) -> StringUtils.isNumeric(event3.getMessageContent().getPlainText()));
        event2.replyBlocking("number: " + event2.getMessageContent().getPlainText());
    }
```
再请问一下，这里抛出的自定义该如何捕获呢？我的想法是应该可以用全局异常捕获，会有别的方法吗？

**Reply by ForteScarlet** [reply]
cancel的异常在对应receiver处接收时可以得到，如果是阻塞的就用try-catch，如果是异步的就使用 CompletableFuture 对应的异常处理API即可

**Reply by YOLO-MIC** [reply]
感谢
